viewer/viewer.js

```diff
@@ -3580,6 +3585,204 @@ app.post('/updateUser/:userId', checkToken, function(req, res) {
   });
 });
 
+app.post('/uploadFragment', checkWebEnabled, fragmentUpload, function(req, res) {
+  Db.getUser(req.user.userId, function(err, user) {
+    if (err || !user.found) {
+      return error("Unknown user");
+    }
+    
+    var fragmentName = req.body.fragmentName;
+    if (fragmentName.length < 3 && fragmentName.length > 20) {
+      return error("Invalid fragment name.");
+    }
+    
+    if (req.files.fragment[0].originalname.indexOf(' ') !== -1) {
+      return error("Fragment filename must not contain spaces.");
+    }
+    
+    user = user._source;
+    var file = req.files.fragment[0];
+    var fragment = {
+      body: file.buffer.toString(),
+      name: fragmentName,
+      type: req.body.fragmentType,
+      filename: file.originalname
+    };
+    if (typeof user.fragments !== 'object') {
+      user.fragments = {};
+    }
+    user.fragments[fragment.filename] = fragment;
```

Comment:
I went with filename because that could be more unique, I would guess that session/query fragment names could clash more often than filenames, although that's pure estimation. Also, it can be used, in JavaScript you can have an array with a key which has a dot in it, as long as you use the bracket syntax to access it.